peer_heap.js

```diff
@@ -20,89 +20,88 @@
 
 'use strict';
 
+var assert = require('assert');
+
 module.exports = PeerHeap;
 
 // A max-score (pre-computed) for peer selection
 
-function PeerHeap() {
+function PeerHeap(random) {
     var self = this;
 
+    self.random = random || Math.random;
+    assert(typeof self.random === 'function', 'PeerHeap expected random fn');
+
     self.array = [];
     // TODO: worth it to keep a tail free list like TimeHeap?
     // self.end = 0;
     self._stack = [];
 }
 
-PeerHeap.prototype.choose = function choose(threshold, filter) {
-    var self = this;
-
-    if (!self.array.length) {
-        return null;
-    }
-
-    var el;
-    if (filter) {
-        el = self._chooseFilteredEl(threshold, filter);
-    } else {
-        el = self._chooseEl(threshold);
-    }
-
-    if (!el) {
-        return null;
-    }
-
-    return el.peer;
-
-};
-
-PeerHeap.prototype._chooseEl = function _chooseEl(threshold) {
+PeerHeap.prototype.chooseWeightedRandom = function chooseWeightedRandom(threshold, filter) {
     var self = this;
 
-    var el = self.array[0];
-    if (el.score <= threshold) { // TODO: why inclusive?
-        return null;
-    }
-
-    return el;
-};
+    // We select into this set as long as we have ranges that overlap. Then,
+    // we introduce randomness into the set and choose
+    var seedPeer = self.array[0].peer;
+    var seedRange = seedPeer.pendingWeightedRange();
+    var set = [seedPeer];
+    var minRangeStart = seedRange[0];
+    var i;
 
-PeerHeap.prototype._chooseFilteredEl = function _chooseFilteredEl(threshold, filter) {
```

Comment:
you got rid of code specialization for the unfiltered vs filtered case, and therefore have left performance in the abundantly common (unfiltered) case on the floor.  Please restore a specialized path for unfiltered.